### DISCUSSION TITLE
color adjustment of histology slides

### DISCUSSION THREAD

**Discussion by Rasaa84** [libvips/libvips#3742] [Q&A]
Hi,

I'm looking to modify the colors of histology slides to match a reference image or a specific section. I am wondering how I can do this using pyvips?

**Comment by jcupitt** [comment]
Hi again, there are loads of ways of doing this, you'd need to be more specific about what you want. Do you have some sample images and an expected result?

**Comment by Rasaa84** [comment]
This is the input image:

![1](https://github.com/libvips/libvips/assets/58141551/a9f35eb1-5823-4d40-a389-fd9727b5e598)

I need to change its color to the color of this image:


![2](https://github.com/libvips/libvips/assets/58141551/dff649b8-1a1f-4273-97b4-747fc8dfa2ef)

**Reply by KempWatson** [reply]
What exactly is wrong with the image?

**Comment by jcupitt** [comment]
That looks like a SVS image from openslide compared to an SVS image from Leica's viewer, is that right?

You need openslide4 plus libvips 8.15, it has all this built in (phew).

**Reply by jcupitt** [reply]
For example:

![image](https://github.com/libvips/libvips/assets/580843/8a2eae35-1093-4fdc-92e0-82a822a57282)

That's an SVS slide with and without colour management. Colour management is a new feature of openslide4 and is supported by libvips:

https://github.com/openslide/openslide/releases/tag/v4.0.0

**Comment by Rasaa84** [comment]
Yes, it's right. one is a SVS from Leica's viewer and the other from openslide. Thanks John for your help. I am wondering if I can use color management along with dzsave?

**Reply by jcupitt** [reply]
Sure, eg.:

```
vips icc_transform 230108509#3##HE#_121909.svs[rgb] x.dz srgb
```

**Reply by jcupitt** [reply]
As well as fixing the colour, it's quite a bit quicker too. 8.14 is:

```
$ time vips icc_transform 230108509#3##HE#_121909.svs[rgb] x.dz srgb
real	0m26.713s
user	4m17.907s
sys	1m55.113s
```

And 8.15 is:

```
$ time vips icc_transform 230108509#3##HE#_121909.svs[rgb] x.dz srgb
real	0m16.748s
user	4m45.958s
sys	0m36.508s
```

**Reply by Rasaa84** [reply]
Thank you so much! what if I want to use it in python?

**Reply by jcupitt** [reply]
It's the same API, so:

```python
image = pyvips.Image.new_from_file("x.svs", rgb=True)
image = image.icc_transform("srgb")
image.dzsave("x.dzi")
```

**Reply by Rasaa84** [reply]
Thanks John!

I used the code but the color didn't changed. This is a tile from dzi file (from the input image I posted above)

![image](https://github.com/libvips/libvips/assets/58141551/a2dd4568-9feb-47f7-a7d3-70c2f22a5e83)

**Reply by jcupitt** [reply]
That's right, you need openslide4 and libvips 8.15, it's a new feature.

**Reply by Rasaa84** [reply]
It worked. Thank you so much for your help!

![image](https://github.com/libvips/libvips/assets/58141551/d539ef87-2373-4fbf-a8ab-27a768b9ebd4)

**Reply by Rasaa84** [reply]
Hi John, 

When I run the same code (the following code) inside a docker:

```
image = pyvips.Image.new_from_file("x.svs", rgb=True)
image = image.icc_transform("srgb")
image.dzsave("x.dzi")
```

I get this error:

`pyvips.error.Error: VipsForeignLoadTiffFile does not support optional argument rgb`

I have libvips 8.15 and openslide 4 in it. Do you have any idea?

![image](https://github.com/libvips/libvips/assets/58141551/4103283a-9974-47ca-a4a5-5b5b3aaee7dc)

![image](https://github.com/libvips/libvips/assets/58141551/cfe19fee-d58f-464e-a2b2-7dc6dd77eaa6)

**Reply by jcupitt** [reply]
Hello again, you've built libvips without openslide support. Perhaps it couldn't find the `.pc` file? Did you set `PKG_CONFIG_PATH`?

**Reply by Rasaa84** [reply]
John, thanks for your response. I resolved the issue but got a new error:

`pyvips.error.Error: no such operation dzsave
  VipsOperation: class "dzsave" not found`

however I have included `libgsf-1-dev`. do you have any idea how I can solve it?

As I said, I built it from source within a Dockerfile. I am wondering whether it's necessary to build libvips from source to obtain version 8.15?

**Reply by jcupitt** [reply]
We got rid of libgsf. Remove that and add libarchive instead.

**Reply by Rasaa84** [reply]
It worked, thanks John!

This is list of packages I have added in my Dockerfile:

```
RUN apt-get install -y \
	glib-2.0-dev \
	libheif-dev \
	libexpat-dev \
	librsvg2-dev \
	libpng-dev \
	libgif-dev \
	libjpeg-dev \
	libtiff-dev \
	libexif-dev \
	liblcms2-dev \
	liborc-dev \
	libffi-dev
RUN apt install -y \
    libfftw3-dev \
    libopenexr-dev \
    libarchive-dev \
    liborc-dev \
    libmatio-dev \
    libwebp-dev \
    libexif-dev \
    libcfitsio-dev \
    libpoppler-glib-dev \
    librsvg2-dev \
    libpango1.0-dev \
    libopenjp2-7-dev \
    libimagequant-dev
```

could you please review and inform me if there is a need to include or exclude certain packages?

Thank you so much in advance.

Best,
Rasoul

**Reply by jcupitt** [reply]
- I'd remove orc and add hwy, assuming you have a recent enough version available. 
- You could also swap libpng for libspng.
- You have librsvg and orc in there twice.
- You don't need libgif
- I'd add libcgif, if available

You have untested libraries in there, so you are very vulnerable to attacks. I would not use this libvips to process any user data.

**Reply by Rasaa84** [reply]
Regarding untested libraries, which libraries do you mean specifically? probably I can remove them. 


I have also another question. I have used the same API to fix the color for two different image from different scanners. The outputs are as follows:

![image](https://github.com/libvips/libvips/assets/58141551/fac7cf5b-5467-44a4-a06c-83a2fc258cb9)

As you see the colors are different. The API changes the color of image from Leica (left image) while keeping the color of image from other scanner (right image) unchanged. I'm interested in obtaining uniform output colors. I'm specifically interested in harmonizing the slide stains to match the color of the left image, aiming for a more pinkish tone rather than a purplish one. I know that there are different approaches for stain transformation but I am wondering if libvips provides such possibility.

**Reply by jcupitt** [reply]
> Regarding untested libraries, which libraries do you mean specifically? probably I can remove them.

libmatio is probably the worst -- it hasn't been designed for untrusted input, so it is very easy to attack. Unless you really need to load Matlab save files, I would remove it. 

If you are loading internet data, I would remove all libraries that you don't need.

> I have used the same API to fix the color for two different image from different scanners.

You'd need to look at your set of scanners, see which ones include an ICC profile, and check if openslide supports ICC profiles from that scanner.

But that probably won't be enough to get a good match. Many scanner manufacturers use ICC profiles to make their slides images look "better" (ie. brighter, more contrasty, more colourful), and have no interest in colour accuracy. 

The best solution is to use a scanner target and to calibrate the devices yourself, eg.:

https://www.appliedimage.com/applied-image-announces-nist-traceable-color-transmission-calibration-slide-digital-pathology-imaging/

(just the fist hit on google, there are many of these)

**Reply by Rasaa84** [reply]
John, Thank you so much! Appreciate your help.

**Reply by jcupitt** [reply]
It's pretty easy to use libvips to find a colour correction matrix from a scan of a calibration target, if you end up going that way.
